### PR TITLE
[Update]: Validate identifiers for UAX #31 and COBOL Standard conformance

### PR DESCRIPTION
**Pull request by KTSnowy**
---
## Otterkit COBOL Identifier Profile for UAX #31 and COBOL Standard conformance

The Unicode standard requires us to document how this works, so I'm creating a PR to document all of it.

[UAX31-C1](https://unicode.org/reports/tr31/#C1): An implementation claiming conformance to this specification shall identify the version of this specification.
- As of this PR, Otterkit is using UAX #31 Revision 37 (from Unicode 15.0.0) as the basis for identifier validation.

[UAX31-C2](https://unicode.org/reports/tr31/#C2): An implementation claiming conformance to this specification shall describe which of the following requirements it observes:
- [UAX31-R1-2](https://unicode.org/reports/tr31/#R1-2): Otterkit uses the following profile of [UAX31-R1-1](https://unicode.org/reports/tr31/#R1-1).
```
// Otterkit COBOL Identifier Profile (profile of UAX31-R1-1):
// Identifiers must not have more than 63 characters and are case-insensitive.
// All identifiers that contain non-normalized characters are converted into the Normalization Form NFKC
        
// Identifier    :=  XID_Start (Medial XID_Continue+)* XID_Continue*
// XID_Start     :=  <All characters in the unicode categories Lu, Ll, Lt, Lm, Lo, Nl, basic digits [0-9], and characters with the Other_ID_Start property>
// XID_Continue  :=  <All characters in XID_Start, plus characters in the unicode categories Mn, Mc, Nd, Pc, and characters with the Other_ID_Continue property>
// Medial        :=  <Only the special characters minus sign (HYPHEN-MINUS), underscore (LOW LINE), and the ・ (KATAKANA MIDDLE DOT)>
        
// Characters with the Other_ID_Start and Other_ID_Continue properties can be found here: https://www.unicode.org/Public/15.0.0/ucd/PropList.txt
```
- [UAX31-R1a](https://unicode.org/reports/tr31/#R1a): Restricted Format Characters are not allowed due to security and usability concerns, Otterkit therefore does not observe this requirement.
- [UAX31-R1b](https://unicode.org/reports/tr31/#R1b): Otterkit will maintain backwards compatibility in all future versions of Unicode through the use of grandfathered characters, and both the Other_ID_Start and Other_ID_Continue characters, therefore we *currently* observe this requirement
- [UAX31-R2](https://unicode.org/reports/tr31/#R2): Otterkit does not have Immutable Identifiers due to NFKC form normalization and case-insensitive equality, therefore we do not observe this requirement.
- [UAX31-R3-2](https://unicode.org/reports/tr31/#R3-2): Otterkit uses the following profile of [UAX31-R3-1](https://unicode.org/reports/tr31/#R3-1) which meets the requirements for Standard COBOL;
```
// Pattern_White_Space  := <The COBOL character space, and the period, comma and semicolon when immediately followed by a space>
// Pattern_Syntax       := <Basic letters [A-Za-z], basic digits [0-9] and the following basic special characters [+ - * / = $ , ; . " ' ( ) > < & : _]>

// Note 1: The separator period must only be used to indicate the end of a sentence (For example, statements).

// Note 2: Although the character period is also part of the Pattern_Syntax, sometimes the separator is optional and acts as an extra whitespace instead of required syntax.

// Note 3: The characters comma and semicolon when immediately followed by a space can be used anywhere the character space is used.
```

- [UAX31-R4](https://unicode.org/reports/tr31/#R4): All identifiers that contain non-normalized characters are converted into the Normalization Form NFKC, therefore we observe this requirement.
- [UAX31-R5](https://unicode.org/reports/tr31/#R5): Otterkit technically observes this requirement for simple case folding using C#'s `OrdinalIgnoreCase`, but this does not force the string into uppercase (mutate the string like the Unicode standard requires) and instead the comparisons are only done when adding and retrieving identifiers from the symbol table (this is done to avoid extra string allocations and improve performance), functionally it works as if the identifiers were uppercased.

>https://learn.microsoft.com/en-us/dotnet/api/system.stringcomparer.ordinalignorecase
>
>The [StringComparer](https://learn.microsoft.com/en-us/dotnet/api/system.stringcomparer?view=net-7.0) returned by the [OrdinalIgnoreCase](https://learn.microsoft.com/en-us/dotnet/api/system.stringcomparer.ordinalignorecase?view=net-7.0) property treats the characters in the strings to compare as if they were converted to uppercase using the conventions of the invariant culture, and then performs a simple byte comparison that is independent of language. This is most appropriate when comparing strings that are generated programmatically or when comparing case-insensitive resources such as paths and filenames.

- [UAX31-R6](https://unicode.org/reports/tr31/#R6): All identifiers that contain non-normalized characters are converted into the Normalization Form NFKC, which in theory means that no characters should be excluded from normalization (but are not forced into performing normalization if they are already normalized), therefore we observe this requirement.
- [UAX31-R7](https://unicode.org/reports/tr31/#R7): Otterkit technically observes this requirement for simple case folding using C#'s `OrdinalIgnoreCase`, no characters are excluded from case-insensitive comparisons. 
- [UAX31-R8](https://unicode.org/reports/tr31/#R8): The COBOL Standard does not allow hashtag characters at the start of identifiers, therefore we do not observe this requirement.

---

### CODE REVIEW

**Review by gabrielesilinic — APPROVED** [review]
I trust you, you are the main dev, also i don't know really about this specific topic, so if you mess up it's your fault, i could never know, i approve just to allow you to merge

**Comment by GitMensch** [comment]
As I was asked to comment...

For the PR itself: that definitely includes a lot of refactoring, in my opinion this should be done up-front and separately, because it is hard to follow what this PR is about and actually _wants_ to adjust.

For the identifiers (which I _think_ is what this is about) - we take the following approach in GnuCOBOL:

* "words" that are reserved - according to the current dialect / reserved words configuration (see `>> COBOL-WORDS`) are never identifiers
* "integers" that are a "label" (so more or less prefixed by a dot and followed by one) - are identifiers (necessary for label names)
* words consisting of _single-byte_ occurrences matching `[A-Z0-9\x80-\xFF]([_A-Z0-9\x80-\xFF-]*[A-Z0-9\x80-\xFF]+)?` are identifiers (UTF16 and UTF32 with its x00 therefore won't work at all, but most single-byte encodings and several words in UTF8 do work)

**Comment by KTSnowy** [comment]
Hey @GitMensch, sorry I should have made another commit for the refactoring.

The main part of this PR is here in the [TryValidateIdentifier](https://github.com/otterkit/otterkit/pull/14/files#diff-d4388c8a28072c107d017a3fb78e2f27ece4e812d2fb9696d8e83d5a526afb7bR11) method ([src/OtterkitLexer/Token.Methods.cs](https://github.com/otterkit/otterkit/pull/14/files#diff-d4388c8a28072c107d017a3fb78e2f27ece4e812d2fb9696d8e83d5a526afb7b)), this is where we're validating all the allowed XID_Start and XID_Continue characters.

> "integers" that are a "label" (so more or less prefixed by a dot and followed by one) - are identifiers (necessary for label names)

Yeah I was wondering about that as well, because the standard allows user-defined paragraphs like "001" or "02" which alone would be classified as a numeric literal. I haven't thought of checking for a prefix and suffix separator period but it makes sense, I'll update the lexer to take that into account when classifying those.

> words consisting of single-byte occurrences matching `[A-Z0-9\x80-\xFF]([_A-Z0-9\x80-\xFF-]*[A-Z0-9\x80-\xFF]+)?` are identifiers (UTF16 and UTF32 with its x00 therefore won't work at all, but most single-byte encodings and several words in UTF8 do work)

Does GnuCOBOL also do NFKC normalization? For example the Unicode character `ﬁ` would be normalized into two characters, `f` and `i` before going through the character validation.

**Comment by GitMensch** [comment]
> Hey @GitMensch, sorry I should have made another commit for the refactoring.

You still could undo the last commit then split into two, then force-push.... (doing so before the merge or even better before the PR would have been best, of course). But that's not my vcs history so...

> Does GnuCOBOL also do NFKC normalization?

No... because we only handle "single character input" - multiple bytes are split and then only accepted if they would be "valid single-byte characters".

GnuCOBOL will very likely add libiconv to handle also input data with UTF16/UTF32 but will do that by internally converting it to UTF8 with the same "either valid or not" approach. Of course we _may_ handle this different later but in my point of view this should only be done upon explicit request / need.
